app/Http/Controllers/Backend/Social/AbstractTwitterController.php

```diff
@@ -0,0 +1,150 @@
+        } else {
+            return new TwitterController();
+        }
+    }
+
+    /**
+     * Function to create a user with a login-provider.
+     * If logged in, the user will have the login-provider added.
+     * If a user with corresponding login-provider already exists, it will be returned.
+     *
+     * @param SocialiteUser $socialiteUser
+     *
+     * @return User model
+     */
+    public static function getUserFromSocialite(SocialiteUser $socialiteUser): User {
+        $socialProfile = SocialLoginProfile::where('twitter_id', $socialiteUser->id)->first();
+
+        if ($socialProfile !== null) {
+            self::updateToken($socialProfile->user, $socialiteUser);
+            return $socialProfile->user;
+        }
+
+        if (auth()->check()) {
+            self::updateToken(auth()->user(), $socialiteUser);
+            return auth()->user();
+        }
+        return self::createUser($socialiteUser);
+    }
+
+    private static function createUser(SocialiteUser $socialiteUser): User {
+        $user = User::create([
+                                 'name'     => SocialController::getDisplayName($socialiteUser),
+                                 'username' => SocialController::getUniqueUsername($socialiteUser->getNickname()),
+                             ]);
+        self::updateToken($user, $socialiteUser);
+        return $user;
+    }
+
+    private static function updateToken(User $user, SocialiteUser $socialiteUser): void {
+        $user->socialProfile->update([
+                                         'twitter_id'               => $socialiteUser->id,
+                                         'twitter_token'            => $socialiteUser->token,
+                                         'twitter_refresh_token'    => $socialiteUser->refreshToken,
+                                         'twitter_token_expires_at' => Date::now()->add('second', $socialiteUser->expiresIn)
+                                     ]);
+    }
+
+
+    /**
+     * @param Status $status
+     * @param string $socialText
+     *
+     * @return int
+     * @throws NotConnectedException
+     * @throws TweetNotSendException
+     */
+    public abstract function postTweet(Status $status, string $socialText): int;
+
+    /**
+     * @throws NotConnectedException
+     */
+    public static function postStatus(Status $status) {
+        $controller = self::forUser($status->user);
+
+        try {
+            $socialText = self::generateFullSocialText($status);
+            $tweetId    = $controller->postTweet($status, $socialText);
+            $status->update(['tweet_id' => $tweetId]);
+            Log::info("Posted on Twitter: " . $socialText);
+        } catch (NotConnectedException $exception) {
+            throw $exception;
+        } catch (TweetNotSendException $exception) {
+            $status->user->notify(new TwitterNotSent($exception->getStatusCode(), $exception->getStatus()));
```

Comment:
Removes the need to carry the status in the exception.

```suggestion
            $status->user->notify(new TwitterNotSent($exception->getStatusCode(), $status));
```